src/main/scala/io/iohk/ethereum/blockchain/sync/SyncStateScheduler.scala

```diff
@@ -0,0 +1,411 @@
+    new SyncStateScheduler(blockchain)
+  }
+
+  final case class StateNodeRequest(
+      nodeHash: ByteString,
+      resolvedData: Option[ByteString],
+      requestType: RequestType,
+      parents: Seq[ByteString],
+      nodeDepth: Int,
+      dependencies: Int
+  ) {
+    def isNodeRequest: Boolean = requestType match {
+      case _: CodeRequest => false
+      case _: NodeRequest => true
+    }
+  }
+
+  private val stateNodeRequestComparator = new Comparator[StateNodeRequest] {
+    override def compare(o1: StateNodeRequest, o2: StateNodeRequest): Int = {
+      o2.nodeDepth compare o1.nodeDepth
+    }
+  }
+
+  implicit class Tuple2Ops[A, B](o: io.vavr.Tuple2[A, B]) {
+    def asScala(): (A, B) = (o._1, o._2)
+  }
+
+  final case class SyncResponse(hash: ByteString, data: ByteString)
+
+  case class SchedulerState(
+      activeRequest: Map[ByteString, StateNodeRequest],
+      queue: PriorityQueue[StateNodeRequest],
+      memBatch: Map[ByteString, (ByteString, RequestType)]
+  ) {
+
+    def schedule(request: StateNodeRequest): SchedulerState = {
+      activeRequest.get(request.nodeHash) match {
+        case Some(oldRequest) =>
+          copy(activeRequest + (request.nodeHash -> oldRequest.copy(parents = oldRequest.parents ++ request.parents)))
+
+        case None =>
+          copy(activeRequest + (request.nodeHash -> request), queue.enqueue(request))
+      }
+    }
+
+    def getMissingHashes(max: Int): (List[ByteString], SchedulerState) = {
+      def go(
+          currentQueue: PriorityQueue[StateNodeRequest],
+          remaining: Int,
+          got: List[ByteString]
+      ): (PriorityQueue[StateNodeRequest], List[ByteString]) = {
+        if (remaining == 0) {
+          (currentQueue, got.reverse)
+        } else if (currentQueue.isEmpty) {
+          (currentQueue, got.reverse)
+        } else {
+          val (elem, newQueue) = currentQueue.dequeue().asScala()
+          go(newQueue, remaining - 1, elem.nodeHash :: got)
+        }
+      }
+
+      val (newQueue, elements) = go(queue, max, List.empty)
+      (elements, copy(queue = newQueue))
+    }
+
+    def getAllMissingHashes: (List[ByteString], SchedulerState) = getMissingHashes(queue.size())
+
+    def numberOfPendingRequests: Int = activeRequest.size
+
+    def getPendingRequestByHash(hash: ByteString): Option[StateNodeRequest] = activeRequest.get(hash)
+
+    def numberOfMissingHashes: Int = queue.size()
+
+    def commit(request: StateNodeRequest): SchedulerState = {
+      @tailrec
+      def go(
+          currentRequests: Map[ByteString, StateNodeRequest],
+          currentBatch: Map[ByteString, (ByteString, RequestType)],
+          parentsToCheck: Seq[ByteString]
+      ): (Map[ByteString, StateNodeRequest], Map[ByteString, (ByteString, RequestType)]) = {
+        if (parentsToCheck.isEmpty) {
+          (currentRequests, currentBatch)
+        } else {
+          val parent = parentsToCheck.head
```

Comment:
WDYT about adding some meaningful exception here? eg. `val parent = parentsToCheck.headOption.getOrElse(throw new IllegalStateException("Critical exception. Cannot find parent"))`

Reply:
it sound like good idea, it will be instantly known that some invariants have been broken